das/daser.go

```diff
@@ -0,0 +1,84 @@
+package das
+
+import (
+	"context"
+	"fmt"
+
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/celestiaorg/celestia-node/service/header"
+	"github.com/celestiaorg/celestia-node/service/share"
+)
+
+var log = logging.Logger("das")
+
+// TODO(@Wondertan): Initialization and warm-up.
+// TODO(@Wondertan): Start and Stop is better be thread-safe.
+// TODO(@Wondertan): Add some testing
+// DASer continuously validates availability of data committed to headers.
+type DASer struct {
+	da   share.Availability
+	hsub header.Subscriber
+
+	cancel context.CancelFunc
+}
+
+// NewDASer creates a new DASer.
+func NewDASer(da share.Availability, hsub header.Subscriber) *DASer {
+	return &DASer{
+		da:   da,
+		hsub: hsub,
+	}
+}
+
+// Start initiates subscription for new Headers and spawns a sampling routine.
+func (d *DASer) Start() error {
+	if d.cancel != nil {
+		return fmt.Errorf("da: DASer already started")
+	}
+
+	sub, err := d.hsub.Subscribe()
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go d.sampling(ctx, sub)
+	d.cancel = cancel
+	return nil
+}
+
+// Stop stops sampling.
+func (d *DASer) Stop() error {
+	if d.cancel == nil {
+		return fmt.Errorf("da: DASer already stopped")
+	}
+
+	d.cancel()
+	d.cancel = nil
+	return nil
+}
+
+// sampling validates availability for each Header received from header subscription.
+func (d *DASer) sampling(ctx context.Context, sub header.Subscription) {
+	defer sub.Cancel()
+	for {
+		h, err := sub.NextHeader(ctx)
+		if err != nil {
+			if err == context.Canceled {
+				return
+			}
+			log.Errorw("daser failed to get next header", "err", err)
+			return
+		}
+
+		err = d.da.SharesAvailable(ctx, h.DAH)
```

Comment:
Do we have an issue open somewhere for error handling when things like `SharesAvailable` fail?

Reply:
While we are in dev/test nets environment I would just:
* Retry
* Persist failed retries with #180 

In a real environment, we should probably go with #84. As SharesAvailable should somehow work with BadEncodingFraudProofs at a later stage.